### PR TITLE
Add typed events for all user hooks and legacy events

### PR DESCRIPTION
**Pull request by ChristophWurst**
# Update

see https://github.com/nextcloud/server/pull/18348

---

* Add typed events
* Emit events where we emitted legacy events and/or hooks
* Catch new events and re-emit data as legacy events and/or hooks

-> existing code should continue to work, but new code can use the typed events with 18+

- [x] Session https://docs.nextcloud.com/server/stable/developer_manual/app/hooks.html#session
  - [x] preSetPassword (\OC\User\User $user, string $password, string $recoverPassword)
  - [x] postSetPassword (\OC\User\User $user, string $password, string $recoverPassword)
  - [x] changeUser (\OC\User\User $user, string $feature, string $value)
  - [x] preDelete (\OC\User\User $user)
  - [x] postDelete (\OC\User\User $user)
  - [x] preCreateUser (string $uid, string $password)
  - [x] postCreateUser (\OC\User\User $user)
  - [x] preLogin (string $user, string $password)
  - [x] postLogin (\OC\User\User $user, string $password)
  - [x] logout ()
- [x] UserManager
  - [x] preSetPassword (\OC\User\User $user, string $password, string $recoverPassword)
  - [x] postSetPassword (\OC\User\User $user, string $password, string $recoverPassword)
  - [x] preDelete (\OC\User\User $user)
  - [x] postDelete (\OC\User\User $user)
  - [x] preCreateUser (string $uid, string $password)
  - [x] postCreateUser (\OC\User\User $user, string $password)

Note: I don't get the diff between `Session` and `UserManager` hooks as we always emitted on the manager.

### CODE REVIEW

File: lib/composer/composer/autoload_static.php [hunk 1]
```diff
@@ -479,9 +479,19 @@ class ComposerStaticInit53792487c5a8370acc0b06b1a864ff4c
         'OCP\\User\\Backend\\IProvideAvatarBackend' => __DIR__ . '/../../..' . '/lib/public/User/Backend/IProvideAvatarBackend.php',
         'OCP\\User\\Backend\\ISetDisplayNameBackend' => __DIR__ . '/../../..' . '/lib/public/User/Backend/ISetDisplayNameBackend.php',
         'OCP\\User\\Backend\\ISetPasswordBackend' => __DIR__ . '/../../..' . '/lib/public/User/Backend/ISetPasswordBackend.php',
-        'OCP\\User\\Events\\CreateUserEvent' => __DIR__ . '/../../..' . '/lib/public/User/Events/CreateUserEvent.php',
-        'OCP\\User\\Events\\PostLoginEvent' => __DIR__ . '/../../..' . '/lib/public/User/Events/PostLoginEvent.php',
+        'OCP\\User\\Events\\BeforePasswordUpdatedEvent' => __DIR__ . '/../../..' . '/lib/public/User/Events/BeforePasswordUpdatedEvent.php',
+        'OCP\\User\\Events\\BeforeUserCreatedEvent' => __DIR__ . '/../../..' . '/lib/public/User/Events/BeforeUserCreatedEvent.php',
```
Comment (kesselb) [inline lib/composer/composer/autoload_static.php]: I find `BeforeUserCreatedEvent`, `BeforeUserDeletedEvent`, etc confusing. Isn't that the wrong tense? :confused:

Comment (ChristophWurst) [reply lib/composer/composer/autoload_static.php]: Is it? "Before the user is deleted (by the system)"? I see the same tense used here https://symfony.com/doc/4.3/components/event_dispatcher.html#naming-conventions

**Review by rullzer — APPROVED** [review]
quick :smoking:-test
looking good. Lets do this :rocket:

File: lib/composer/composer/autoload_static.php (hunk 1 above)
Comment (kesselb) [reply lib/composer/composer/autoload_static.php]: OK

**Comment by ChristophWurst** [comment]
rebased to resolve conflicts

**Comment by ChristophWurst** [comment]
updated to fix tests

**Review by skjnldsv — APPROVED** [review]
Houston, We have a liftoff :rocket:

**Comment by MichaIng** [comment]
This solves #18331, right?

**Comment by ChristophWurst** [comment]
> This solves #18331, right?

Nope. The uses of the old event are still a problem in some cases. I'll try to have a look at that as well. The problem is that we can't migrate those without introducing a breaking change :cry:

**Comment by MichaIng** [comment]
@ChristophWurst 
Okay, if it does not cause issues now, hence does not require urgent action, it is probably best to omit those logs, to not flood log files?

**Comment by ChristophWurst** [comment]
Those logs are fine. You only see them with log minimum log levels set.

We want them to be visible for app developers so they take action timely. On production instances these lines won't show up.

**Comment by MichaIng** [comment]
@ChristophWurst 
It is currently log level 1 (info) at least, not "minimum" log level 0 (debug). Does it make sense to report it to re individual apps repos then? Not sure how quickly different developers test NC betas 😉.

**Comment by ChristophWurst** [comment]
It's not just about the apps. As long as we emit the legacy event for BC reasons this info log will be shown.

**Comment by ChristophWurst** [comment]
Can't make CI happy here. https://github.com/nextcloud/server/pull/18348 is a simplified version for 18. Cleanup has to follow later.

**Comment by gary-kim** [comment]
@ChristophWurst I think we can close this now?

**Pull request closed without merging**